libraries/eosiolib/core/eosio/crypto_bls_ext.hpp

```diff
@@ -149,13 +326,271 @@ namespace eosio {
+        uint8_t b_i[SHA256HashSize + 1] = { 0, };
+        for (unsigned i = 1; i <= ell; ++i) {
+            for (unsigned j = 0; j < SHA256HashSize; ++j) {
+                b_i[j] = b_0[j] ^ b_i[j];
+            }
+            b_i[SHA256HashSize] = i;
+            bls12_381::sha256 s;
+            s.update(b_i, SHA256HashSize + 1);
+            s.update(dst, dst_len);
+            s.update(dstlen_str, 1);
+            s.digest(b_i);
+            const int rem_after = buf_len - i * SHA256HashSize;
+            const int copy_len = SHA256HashSize + (rem_after < 0 ? rem_after : 0);
+            memcpy(buf + (i - 1) * SHA256HashSize, b_i, copy_len);
+        }
+    }
+    bls_s scalar_fromBE(const bls_s& in) {
+        std::array<uint64_t, 8> out;
+        for(uint64_t i = 0; i < 8; i++) {
+            uint64_t temp;
+            memcpy(&temp, &in[in.size() - i*8 - 8], sizeof(uint64_t));
+            out[i] = htobe64(temp);
+        }
+        return reinterpret_cast<bls_s&&>(std::move(out));
+    }
+    void g2_fromMessage(const bls_g1_affine& msg, const std::string& dst, bls_g2& res) {
+
+        std::array<bls_s, 4> buf;
+        xmd_sh256(buf.data()->data(), sizeof(buf), msg.data(), msg.size(), dst.data(), dst.length());
+
+        bls_s k;
+        bls_fp2 t;
+        bls_g2 p, q;
+
+        k = scalar_fromBE(buf[0]);
+        bls_fp_mod(k, t[0]);
+        k = scalar_fromBE(buf[1]);
+        bls_fp_mod(k, t[1]);
+
+        bls_g2_map(t, p);
+
+        k = scalar_fromBE(buf[2]);
+        bls_fp_mod(k, t[0]);
+        k = scalar_fromBE(buf[3]);
+        bls_fp_mod(k, t[1]);
+
+        bls_g2_map(t, q);
+        bls_g2_add(p, q, res);
+    }
+}
+
+    inline std::string bls_g1_affine_to_base64(const bls_g1_affine& g1) {
+        return detail::bls_type_to_base64<bls_g1_affine, detail::bls_public_key_prefix>(g1);
+    }
+    inline bls_g1_affine bls_base64_to_g1_affine(const char* data, size_t size) {
+        return detail::bls_base64_to_type<bls_g1_affine, detail::bls_public_key_prefix>(data, size);
+    }
+    inline std::string bls_sig_to_base64_affine(const bls_g2_affine& g2) {
+        return detail::bls_type_to_base64<bls_g2_affine, detail::bls_signature_prefix>(g2);
+    }
+    inline bls_g2_affine bls_base64_to_sig_affine(const char* data, size_t size) {
+        return detail::bls_base64_to_type<bls_g2_affine, detail::bls_signature_prefix>(data, size);
+    }
+
+    // pubkey and signature are assumed to be in RAW affine little-endian bytes
+    bool bls_pop_verify(const bls_g1_affine& pubkey, const bls_g2_affine& signature_proof) {
+        using namespace detail;
+
+        bls_g1 g1_points[2] = {0};
+        bls_g2 g2_points[2] = {0};
+
+        // add z coordinate (R1) to pubkey and signature_proof
+        std::vector<uint8_t> pubkey_ex(sizeof(bls_g1_affine) + R1.size(), 0);
+        auto insert_it = std::copy(pubkey.begin(), pubkey.end(), pubkey_ex.begin());
+        std::copy(R1.begin(), R1.end(), insert_it);
```

Comment:
Pull this (or something like it) out to the following helper function which should also be publicly accessible:

```cpp
inline bls_g1 g1_affine_to_jacobian(const bls_g1_affine& g1);
```